Modules/fact_manager.py

```diff
@@ -0,0 +1,336 @@
+"""
+fact_manager.py - Fact Management
+
+Provides a context manager for manipulating, retrieving, and storing facts.
+This class requires database.DatabaseManager and Modules.fact.
+
+Copyright (c) 2018 The Fuel Rat Mischief,
+All rights reserved.
+
+Licensed under the BSD 3-Clause License.
+
+See LICENSE.md
+"""
+import datetime
+import logging
+from datetime import timezone
+
+import psycopg2
+from psycopg2 import sql
+
+
+from Modules.fact import Fact
+from database import DatabaseManager
+
+log = logging.getLogger(f"mecha.{__name__}")
+
+
+class FactManager(DatabaseManager):
+
+    def __init__(self, fact_table="fact2", fact_log="fact_transaction"):
+        if not isinstance(fact_table, str):
+            raise TypeError("fact table must be a string.")
+
+        if not isinstance(fact_log, str):
+            raise TypeError("fact log must be a string.")
+
+        self._FACT_TABLE = fact_table
+        self._FACT_LOG = fact_log
+
+        # Proclaim loudly into the void that we are loaded.
+        log.info("Fact Manager Initialized.")
+        super().__init__()
+
+    async def add(self, fact: Fact):
+        """
+        Adds a new fact to the database.  This will result in a ProgrammingError being thrown
+        if a fact violates the (name, lang) Primary Key relationship.
+
+        VERIFY IF THE FACT EXISTS WITH self.EXISTS() FIRST.
+
+        >>> if self.exists("test", "en"):
+        ...     await self.add(NewFact)
+        ... else:
+        ...     # Abandon Ship!
+        ...     raise TypeError("Attempted commit on incomplete Fact.")
+
+        Aliases cannot be inserted this way, at this time.  None is forced for fact.aliases.
+        """
+        # Don't judge this SQL, its the most compact way to do it. (Thanks PEP8)
+        add_query = sql.SQL(f"INSERT INTO {self._FACT_TABLE} "
+                            f"VALUES ( %s, %s, %s, %s, %s, %s, %s, %s )")
+
+        try:
+            if fact.complete:
+                # No tuple to unpack here, so assign manually:
+                add_values = (fact.name, fact.lang, fact.message, None,
+                              fact.author, fact.edited, fact.editedby, fact.mfd)
+
+                # run INSERT query
+                await self.query(add_query, add_values)
+            else:
+                # Lets handle this TypeError here, because if .complete is False, we didn't
+                # get enough information to create it without violating NOT NULL fields.
+                raise TypeError("Attempted commit on incomplete Fact.")
+
+        except (psycopg2.DatabaseError, psycopg2.IntegrityError) as error:
+            # Database is not available, or fact already exists and wasn't checked.
+            # Raise error, generating graceful cheese error and log the exception.
+            log.exception(f"Unable to add fact '{fact.name}!", error)
```

Comment:
We talked about this, `log.exception` **should not** be the exception passed as an argument!

Reply:
fixed in 83935a7